proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+# Background
+
+Cryptography is an essential part of modern computing and information security, but it requires care, nuance, and knowledge to properly implement, and is extremely easy to mess up. It covers a diverse range of functions, with individual algorithms often having their own particular quirks that must be acknowledged to properly and safely use. Failure to properly implement cryptography carries the risk of severely compromising the security of any system in which it is used. Manually implementing common cryptographic primitives such as hashes, ciphers, digital signatures, big-integer arithmetic is thus considered extremely unwise, both in terms of security and performance.
+
+Cryptography is even harder for functional programming languages. Properly implementing techniques such as sanitizing memory, performing operations in constant time, are already difficult enough in languages such as C, and in a lazy functional language it is made even more difficult by such things as the potential for references to be kept alive within thunks. As a result, the space of functional cryptography is relatively under-explored and under-developed, having favored other languages where such things are easier to manage.
+
+However, if we accept this downside as a part of a trade-off, and explore the ways that functional programming is beneficial to cryptography (monads, type systems, referential transparency), there is then also the potential for functional cryptography to be made *easier*. The functional machinery of Haskell is well-suited for adding important contextual control to cryptography, and can help prevent or eliminate many issues and errors in ways that languages such as C cannot. I believe that there is room for significant improvement, given a directed effort to develop the space.
+
+# Problem Statement
+
+Cryptography in Haskell lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and exposes end users to significant risk in the event of a lapse in security. Some companies have built their own solution to this, but there is no community-driven, community-owned solution.
+
+Cryptography in Haskell is also fragile and outdated, having seen considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. This has placed the long-term stability of important libraries such as `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented.
```

Comment:
```suggestion
Haskell cryptographic libraries are also presently fragile and outdated, having seen considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. This has placed the long-term stability of important libraries such as `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented.
```
What does "fragile" mean here? I'd expect it to mean that the library doesn't work well (e.g. segfaults when given slightly bogus input), but it sounds to me like you're instead talking about the development process for the libraries.

Reply:
That is accurate. I have two factors in mind when I say 'fragile':

1) Difficulty in ensuring adoption of new libraries when old ones are deprecated - the cryptography ecosystem in haskell suffers from this extensively as the organization of modules and libraries has changed over time while code has moved from `cryptohash / cryptocipher` to `crypton` and `cryptonite`. As a result, there are multiple cryptography libraries with almost the same interface, because of a shared lineage, and it is not always clear which should be used preferentially. This promotes rot as dependents of unmaintained libraries finally stop working.

2) These libraries have highly concrete implementations / bad abstractions / interfaces which are strongly influenced by their implementation, but have propagated due to shared lineage. While the interfaces themselves haven't really changed much, the same code has been moved from library to library, in part because it violates the abstraction barrier such that if we change backends, it could require breaking the interface - which makes the logic itself brittle. Yet, there is still so much built on it, which also must change in order to move forward.